File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOBloodGlucoseModel.kt
package com.idosmart.model

import com.google.gson.GsonBuilder
import com.google.gson.annotations.SerializedName

/**
 * IDOBloodGlucoseModel
 */
internal class IDOBloodGlucoseModel(
    operate: Int,
    currentInfoItem: IDOBloodGlucoseCurrentInfo? = null,
    statisticsInfoItem: IDOBloodGlucoseStatisticsInfo? = null,
    historyInfoItem: IDOBloodGlucoseHistoryDataInfo? = null

) : IDOBaseModel {
    @SerializedName("version")
    private var version: Int = 0 // 默认值 0

    @SerializedName("operate")
    var operate: Int = operate

    @SerializedName("current_info_count")
    private var currentInfoCount: Int = if (currentInfoItem != null) 1 else 0

    /**
     * 血糖统计数据count 通常为1
     */
    @SerializedName("statistics_info_count")
    private var statisticsInfoCount: Int = if (statisticsInfoItem != null) 1 else 0

    /**
     * 血糖CGM历史数据count 通常为1
     */
    @SerializedName("history_info_count")
    private var historyInfoCount: Int = if (historyInfoItem != null) 1 else 0

    @SerializedName("current_info")
    var currentInfoItem: IDOBloodGlucoseCurrentInfo? = currentInfoItem

    @SerializedName("statistics_info")
    var statisticsInfoItem: IDOBloodGlucoseStatisticsInfo? = statisticsInfoItem

    @SerializedName("history_info")
    var historyInfoItem: IDOBloodGlucoseHistoryDataInfo? = historyInfoItem

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


/**
 * IDOBloodGlucoseCurrentInfo
 */
open class IDOBloodGlucoseCurrentInfo(
    lastUtcDate: Long,
    lastGlucoseVal: Int = 0xFFFF,
    targetUnit: Int,
    trendVal: Int,
    sensorWarmUpTime: Int,
    sensorStatus: Int,
    normalGlucoseValMax: Int,
    normalGlucoseValMin: Int
) : IDOBaseModel {
    /**
     * 最近一次测量的时间戳
     */
    @SerializedName("last_utc_date")
    var lastUtcDate: Long = lastUtcDate

    /**
     * 最近一次测量的值 100倍
     */
    @SerializedName("last_glucose_val")
    var lastGlucoseVal: Int = lastGlucoseVal

    /**
     * 血糖单位 1: mmol/L 2: mg/DL
     */
    @SerializedName("targetUnit")
    var targetUnit: Int = targetUnit

    /**
     * 血糖趋势
     * 0：无效
     * 1: rapidlyfalling
     * 2: falling
     * 3: slowlyfalling
     * 4: steady
     * 5: slowlyrising
     * 6: rising
     * 7: rapidlyrising
     */
    @SerializedName("trend_val")
    var trendVal: Int = trendVal

    /**
     * 传感器预热时间（单位：分钟）
     */
    @SerializedName("sensor_warm_up_time")
    var sensorWarmUpTime: Int = sensorWarmUpTime

    /**
     * 传感器状态
     * 0：无效
     * 1:正常，在这个状态 血糖数值，trend才有用
     * 2:sensor稳定中
     * 3:sensor错误
     * 4:sensor替换
     * 5:确认新 sensor
     * 6:预热中
     * 7:sensor 过期
     * 8:数据无效
     * 9:sensor低电量
     */
    @SerializedName("sensor_status")
    var sensorStatus: Int = sensorStatus

    /**
     * 血糖正常值范围最大值(100倍)
     */
    @SerializedName("normal_glucose_val_max")
    var normalGlucoseValMax: Int = normalGlucoseValMax

    /**
     * 血糖正常值范围最小值(100倍)
     */
    @SerializedName("normal_glucose_val_min")
    var normalGlucoseValMin: Int = normalGlucoseValMin

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


/**
 * IDOBloodGlucoseStatisticsItem
 */
open class IDOBloodGlucoseStatisticsItem(
    glucoseMax: Int = 0xFFFF,
    glucoseMin: Int = 0xFFFF,
    maxFlu: Int = 0xFFFF,
    ehba1c: Int = 0xFFFF,
    mbg: Int = 0xFFFF,
    sdbg: Int = 0xFFFF,
    cv: Int = 0xFFFF,
    mage: Int = 0xFFFF
) : IDOBaseModel {
    /**
     * 血糖最大值 扩大一百倍
     */
    @SerializedName("glucose_max")
    var glucoseMax: Int = glucoseMax

    /**
     * 血糖最小值 扩大一百倍
     */
    @SerializedName("glucose_min")
    var glucoseMin: Int = glucoseMin

    /**
     * 扩大一百倍
     */
    @SerializedName("maxflu")
    var maxFlu: Int = maxFlu

    /**
     * 扩大一百倍
     */
    @SerializedName("ehba1c")
    var ehba1c: Int = ehba1c

    /**
     * 扩大一百倍
     */
    @SerializedName("mbg")
    var mbg: Int = mbg

    /**
     * 扩大一百倍
     */
    @SerializedName("sdbg")
    var sdbg: Int = sdbg

    /**
     * 扩大一百倍
     */
    @SerializedName("cv")
    var cv: Int = cv

    /**
     * 扩大一百倍
     */
    @SerializedName("mage")
    var mage: Int = mage

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


/**
 * IDOBloodGlucoseStatisticsInfo
 */
open class IDOBloodGlucoseStatisticsInfo(
    targetUnit: Int,
    insulinDose: Int,
    todayItem: IDOBloodGlucoseStatisticsItem? = null,
    weekItem: IDOBloodGlucoseStatisticsItem? = null,
    monthItem: IDOBloodGlucoseStatisticsItem?= null
) : IDOBaseModel {
    /**
     * 血糖单位 1: mmol/L  2: mg/DL
     */
    @SerializedName("targetUnit")
    var targetUnit: Int = targetUnit

    /**
     * 胰岛素注射量 四位小数扩大10000倍
     */
    @SerializedName("insulin_count")
    var insulinDose: Int = insulinDose

    /**
     * 当日统计数据
     */
    @SerializedName("today_item")
    var todayItem: IDOBloodGlucoseStatisticsItem? = todayItem

    /**
     * 本周统计数据
     */
    @SerializedName("week_item")
    var weekItem: IDOBloodGlucoseStatisticsItem? = weekItem

    /**
     * 本月统计数据
     */
    @SerializedName("month_item")
    var monthItem: IDOBloodGlucoseStatisticsItem? = monthItem

    @SerializedName("today_count")
    var todayCount: Int = if (todayItem != null) 1 else 0

    @SerializedName("week_count")
    var weekCount: Int = if (weekItem != null) 1 else 0

    @SerializedName("month_count")
    var monthCount: Int = if (monthItem != null) 1 else 0

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


/**
 * IDOBloodGlucoseHistoryDataInfo
 */
open class IDOBloodGlucoseHistoryDataInfo(
    // 构造函数只接收 dataInfo，其他参数在属性初始化时计算
    dataInfos: List<IDOBloodGlucoseHistoryDataItem>?
) : IDOBaseModel {

    // Swift 构造函数中的计算逻辑被转移到属性初始化中
    /**
     * 总item个数
     */
    @SerializedName("all_items_num")
    var allItemsNum: Int = dataInfos?.size ?: 0 // 对应 self.allItemsNum = dataInfo?.count ?? 0

//    @SerializedName("finish_items_num")
//    var finishItemsNum: Int = 0 // 对应 self.finishItemsNum = 0
//
//    @SerializedName("cur_items_num")
//    var curItemsNum: Int = 1 // 对应 self.curItemsNum = 1
//
//    @SerializedName("history_data_count")
//    var historyDataCount: Int = dataInfo?.size ?: 0 // 对应 self.historyDataCount = dataInfo?.count ?? 0

    /**
     * 历史数据项列表
     */
    @SerializedName("data_infos")
    var dataInfos: List<IDOBloodGlucoseHistoryDataItem>? = dataInfos

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


/**
 * IDOBloodGlucoseHistoryDataItem
 */
open class IDOBloodGlucoseHistoryDataItem(
    date: Long, // 对应 Swift 的 UInt32，使用 Long 以确保无符号 32 位值的安全存储
    glucoseVal: Int,
    targetUnit: Int
) : IDOBaseModel {
    /**
     * 时间
     */
    @SerializedName("date")
    var date: Long = date

    /**
     * 血糖值 扩大一百倍
     */
    @SerializedName("glucose_val")
    var glucoseVal: Int = glucoseVal

    /**
     * 血糖单位 1: mmol/L  2: mg/DL
     */
    @SerializedName("targetUnit")
    var targetUnit: Int = targetUnit

    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOBloodGlucoseModel.swift
//
//  IDOGlucoseCurrentInfo.swift
//  protocol_channel
//
//  Created by hc on 10/27/25.
//

import Foundation

// MARK: - IDOBloodGlucoseModel
@objcMembers
class IDOBloodGlucoseModel: NSObject, IDOBaseModel {
    
    private var version: Int = 0
    
    public var operate: Int
    
    private var currentInfoCount: Int
    
    // 血糖统计数据count 通常为1
    private var statisticsInfoCount: Int
    
    // 血糖CGM历史数据count 通常为1
    private var historyInfoCount: Int
    
    // 预留字节
    //private var data: [Int]
    
    public var currentInfoItem: IDOBloodGlucoseCurrentInfo? = nil
    
    public var statisticsInfoItem: IDOBloodGlucoseStatisticsInfo? = nil
    
    public var historyInfoItem: IDOBloodGlucoseHistoryDataInfo? = nil
    
    
    public init(operate: Int, currentInfoItem: IDOBloodGlucoseCurrentInfo? = nil,  statisticsInfoItem: IDOBloodGlucoseStatisticsInfo? = nil, historyInfoItem: IDOBloodGlucoseHistoryDataInfo? = nil) {
        self.version = 0
        self.operate = operate
        self.currentInfoCount = currentInfoItem != nil ? 1 : 0
        self.statisticsInfoCount = statisticsInfoItem != nil ? 1 : 0
        self.historyInfoCount = historyInfoItem != nil ? 1 : 0
        self.currentInfoItem = currentInfoItem
        self.statisticsInfoItem = statisticsInfoItem
        self.historyInfoItem = historyInfoItem
        //self.data = []
    }
    
    // MARK: - CodingKeys
    public enum CodingKeys: String, CodingKey {
        case version
        case operate
        case currentInfoCount = "current_info_count"
        case statisticsInfoCount = "statistics_info_count"
        case historyInfoCount = "history_info_count"
        //case data
        case currentInfoItem = "current_info"
        case statisticsInfoItem = "statistics_info"
        case historyInfoItem = "history_info"
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            return String(data: jsonData, encoding: .utf8)
        }
        return nil
    }
}



// MARK: - IDOGlucoseCurrentInfo

@objcMembers 
public class IDOBloodGlucoseCurrentInfo: NSObject, IDOBaseModel {
    
    /// 最近一次测量的时间戳
    public var lastUtcDate: UInt32
    
    /// 血糖单位 1: mmol/L 2: mg/DL
    public var targetUnit: Int
    
    /// 血糖趋势
    public var trendVal: Int
    
    /// 传感器状态
    /// 0：无效
    /// 1:正常，在这个状态 血糖数值，trend才有用
    /// 2:sensor稳定中
    /// 3:sensor错误
    /// 4:sensor替换
    /// 5:确认新 sensor
    /// 6:预热中
    /// 7:sensor 过期
    /// 8:数据无效
    /// 9:sensor低电量
    public var sensorStatus: Int

    /// 最近一次测量的值 100倍
    public var lastGlucoseVal: Int = 0xFFFF

    /// 传感器预热时间（单位：分钟）
    public var sensorWarmUpTime: Int
    
    /// 血糖正常值范围最大值(100倍)
    public var normalGlucoseValMax: Int
    
    /// 血糖正常值范围最小值(100倍)
    public var normalGlucoseValMin: Int
    
    public enum CodingKeys: String, CodingKey {
        case lastUtcDate = "last_utc_date"
        case lastGlucoseVal = "last_glucose_val"
        case targetUnit = "targetUnit"
        case trendVal = "trend_val"
        case sensorStatus = "sensor_status"
        case sensorWarmUpTime = "sensor_warm_up_time"
        case normalGlucoseValMax = "normal_glucose_val_max"
        case normalGlucoseValMin = "normal_glucose_val_min"
    }
    
    public init(lastUtcDate: UInt32, lastGlucoseVal: Int = 0xFFFF, targetUnit: Int, trendVal: Int, sensorWarmUpTime: Int, sensorStatus: Int, normalGlucoseValMax: Int, normalGlucoseValMin: Int) {
        self.lastUtcDate = lastUtcDate
        self.lastGlucoseVal = lastGlucoseVal
        self.targetUnit = targetUnit
        self.trendVal = trendVal
        self.sensorStatus = sensorStatus
        self.sensorWarmUpTime = sensorWarmUpTime
        self.normalGlucoseValMax = normalGlucoseValMax
        self.normalGlucoseValMin = normalGlucoseValMin
    }

}


// MARK: - IDOGlucoseStatisticsItem

@objcMembers 
public class IDOBloodGlucoseStatisticsItem: NSObject, IDOBaseModel {
    
    /// 血糖最大值 扩大一百倍
    public var glucoseMax: Int = 0xFFFF
    
    /// 血糖最小值 扩大一百倍
    public var glucoseMin: Int = 0xFFFF
    
    /// 扩大一百倍
    public var maxFlu: Int = 0xFFFF
    
    /// 扩大一百倍
    public var ehba1c: Int = 0xFFFF
    
    /// 扩大一百倍
    public var mbg: Int = 0xFFFF
    
    /// 扩大一百倍
    public var sdbg: Int = 0xFFFF
    
    /// 扩大一百倍
    public var cv: Int = 0xFFFF
    
    /// 扩大一百倍
    public var mage: Int = 0xFFFF
    
    /// 预留字节，长度为 6
    //public var data: [Int]
    
    // MARK: - Initializer (构造方法)
    public init(glucoseMax: Int = 0xFFFF, glucoseMin: Int = 0xFFFF, maxFlu: Int = 0xFFFF, ehba1c: Int = 0xFFFF, mbg: Int = 0xFFFF, sdbg: Int = 0xFFFF, cv: Int = 0xFFFF, mage: Int = 0xFFFF) {
        self.glucoseMax = glucoseMax
        self.glucoseMin = glucoseMin
        self.maxFlu = maxFlu
        self.ehba1c = ehba1c
        self.mbg = mbg
        self.sdbg = sdbg
        self.cv = cv
        self.mage = mage
    }
    
    // MARK: - CodingKeys
    public enum CodingKeys: String, CodingKey {
        case glucoseMax = "glucose_max"
        case glucoseMin = "glucose_min"
        case maxFlu = "maxflu"
        case ehba1c
        case mbg
        case sdbg
        case cv
        case mage
    }
    
}

// MARK: - IDOBloodGlucoseStatisticsInfo

@objcMembers 
public class IDOBloodGlucoseStatisticsInfo: NSObject, IDOBaseModel {
    
    /// 血糖单位 1: mmol/L  2: mg/DL
    public var targetUnit: Int
    
    /// 胰岛素注射量 四位小数扩大10000倍
    public var insulinDose: Int
    
    // 预留字节，长度为 8
    //public var data: [Int]

    private var todayCount: Int
    private var weekCount: Int
    private var monthCount: Int
    
    /// 当日统计数据
    public var todayItem: IDOBloodGlucoseStatisticsItem? = nil
    
    /// 本周统计数据
    public var weekItem: IDOBloodGlucoseStatisticsItem? = nil
    
    /// 本月统计数据
    public var monthItem: IDOBloodGlucoseStatisticsItem? = nil
    
    public init(targetUnit: Int, insulinDose: Int, todayItem: IDOBloodGlucoseStatisticsItem? = nil, weekItem: IDOBloodGlucoseStatisticsItem? = nil, monthItem: IDOBloodGlucoseStatisticsItem? = nil) {
        self.targetUnit = targetUnit
        self.insulinDose = insulinDose
        //self.data = data
        self.todayItem = todayItem
        self.weekItem = weekItem
        self.monthItem = monthItem
        self.todayCount = todayItem != nil ? 1 : 0
        self.weekCount = weekItem != nil ? 1 : 0
        self.monthCount = monthItem != nil ? 1 : 0
    }
    
    public enum CodingKeys: String, CodingKey {
        case targetUnit = "targetUnit"
        case insulinDose = "insulin_count"
        //case data
        case todayItem = "today_item"
        case weekItem = "week_item"
        case monthItem = "month_item"
        case todayCount = "today_count"
        case weekCount = "week_count"
        case monthCount = "month_count"
    }

}


// MARK: - IDOBloodGlucoseHistoryDataInfo
@objcMembers 
public class IDOBloodGlucoseHistoryDataInfo: NSObject, IDOBaseModel {
    
    /// 总item个数
    internal var allItemsNum: Int
    
    /// 已发送/接收的item个数
    //internal var finishItemsNum: Int
    
    /// 当前包item个数 (最大130）
    //internal var curItemsNum: Int
    
    /// 历史血糖个数
    //internal var historyDataCount: Int
    
    /// 历史数据项列表
    public var dataInfos: [IDOBloodGlucoseHistoryDataItem]?
    
    // MARK: - Initializer (构造方法)
    public init(dataInfos: [IDOBloodGlucoseHistoryDataItem]?) {
        self.allItemsNum = dataInfos?.count ?? 0
        //self.finishItemsNum = 0
        //self.curItemsNum = 0
        //self.historyDataCount = dataInfo?.count ?? 0
        self.dataInfos = dataInfos
    }
    
    // MARK: - CodingKeys
    public enum CodingKeys: String, CodingKey {
        case allItemsNum = "all_items_num"
//        case finishItemsNum = "finish_items_num"
//        case curItemsNum = "cur_items_num"
//        case historyDataCount = "history_data_count"
        case dataInfos = "data_infos"
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            return String(data: jsonData, encoding: .utf8)
        }
        return nil
    }
}

// MARK: - IDOGlucoseHistoryDataItem

@objcMembers 
public class IDOBloodGlucoseHistoryDataItem: NSObject, IDOBaseModel {
    
    // 时间
    public var date: UInt32
    
    // 血糖值 扩大一百倍
    public var glucoseVal: Int
    
    // 血糖单位 1: mmol/L  2: mg/DL
    public var targetUnit: Int
    
    public init(date: UInt32, glucoseVal: Int, targetUnit: Int) {
        self.date = date
        self.glucoseVal = glucoseVal
        self.targetUnit = targetUnit
    }
    
    public enum CodingKeys: String, CodingKey {
        case date
        case glucoseVal = "glucose_val"
        case targetUnit = "targetUnit"
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            return String(data: jsonData, encoding: .utf8)
        }
        return nil
    }
}

